specification.md

```diff
@@ -421,6 +421,35 @@ an individual job status callback or not. To remove the callback, set it
 to `null`.
 
 
+<a name="jobexecutor-list"></a>
+```java
+List<String> list()
+```
+
+Return a list of job IDs which are known to this executor instance.  The
+returned list MAY contain IDs of jobs which were not submitted via this
+instance, and MAY NOT contain IDs of jobs which have been submitted by this
+instance, but are finalized and purged already.  IDs for any job which has been
+submitted via this instance and which is not yet in a final state MUST be
+returned.  The returned IDs can be used to (re-)attach a job instance to the job
+via the `executor.attach(id)` call.  This implies that the call SHALL only
+return those IDs to which the callee can attach with under currently used
+authorization.
+
+
+<a name="jobexecutor-attach"></a>
+```java
+Job attach(String jobID)
+```
+
+Return a job which is attached to the backend job identified by the specified
+ID.  The returned job instance MAY have an `UNKNOWN` state until the
+implementation asynchronously obtained information about the actual job state.
+The method MUST raise an `InvalidJobException` if the job ID cannot be used to
+uniquely identify a job.  The `jobID` can either refer to the ID returned by
+`Job.getID()`, which represents the ID used by the J/PSI implementation, or to
+the native job ID used by the LRM backend.
```

Comment:
> The `jobID` can either refer to the ID returned by `Job.getID()`, which represents the ID used by the J/PSI implementation

Is J/PSI assumed to be  stateless (that was an implicit assumption I just realized I was making)?  If so, how do you envision this functionality could be implemented?  If the J/PSI client stores some J/PSI ids to a file, restarts, and then tries to re-attach to the J/PSI ids in that file, how is the J/PSI library/runtime supposed to connect the old J/PSI ids to the underlying LRM's IDs?  Hopefully that question makes sense.

Reply:
Right. We should perhaps also think forward on this particular issue, such as when talking to a remote service, in which case we need to also store and endpoint together with the LRM ID.

Reply:
Throwing this in here as reference points (not advocating one or the other):
  * DRMAA uses the underlying LRM job ID.  A job is created by the equivalent of `executor.submit()` and only returned once it has a job ID
  * SAGA uses the form `[LRM_ID]-[JOB_ID]`, where `LRM_ID` is the url to contact the LRM, and `JOB_ID` is the native job ID in that LRM instance.  SAGA jobs can have an undefined ID until the LRM assigns one to the job,
  
I am a bit worried that having two different job IDs may be confusing.  

Stateless: I did not make that assumption, but it is tempting.  But indeed, it makes consistent mapping to native IDs a bit difficult.  

> ... remote service, in which case we need to also store and endpoint together with the LRM ID.

Our `attach` method is on the Executor, and to get the executor, you would need to know the underlying LRM ID, so the native job IDs would be presumably unique in the scope of that LRM.  So yes, one needs both pieces of information, but it does not necessarily be part of the job ID.

Reply:
> I am a bit worried that having two different job IDs may be confusing.

I share that worry.

> SAGA uses the form [LRM_ID]-[JOB_ID], where LRM_ID is the url to contact the LRM, and JOB_ID is the native job ID in that LRM instance. SAGA jobs can have an undefined ID until the LRM assigns one to the job,

That makes sense.  I would be fine with requiring as an argument to `attach` either `[LRM_ID]` or `[LRM_ID]-[JOB_ID]`.   I have reservations about adding to the standard the ability to just take in a `JOB_ID` without some other stated mechanism in the standard for recovering the underlying `LRM_ID` from a `JOB_ID` after a restart of the J/PSI client.  There are a bunch of ways we could do that, and I think we should open an issue to discuss that further.